File: 86.partition-list.java
/*
 * @lc app=leetcode id=86 lang=java
 *
 * [86] Partition List
 */

// @lc code=start
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode
 * next; ListNode() {} ListNode(int val) { this.val = val; } ListNode(int val,
 * ListNode next) { this.val = val; this.next = next; } }
 */
class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode p = head;
        ListNode small = new ListNode(0), large = new ListNode(0);
        ListNode smallHead = small, largeHead = large;

        while (p != null) {
            if (p.val < x) {
                small.next = p;
                small = small.next;
            } else {
                large.next = p;
                large = large.next;
            }

            p = p.next;
        }

        large.next = null;
        small.next = largeHead.next;

        return smallHead.next;
    }
}
// @lc code=end

File: 86.partition-list.py
#
# @lc app=leetcode id=86 lang=python3
#
# [86] Partition List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def partition(self, head: ListNode, x: int) -> ListNode:

        # before and after are the two pointers used to create two list
        # before_head and after_head are used to save the heads of the two lists.
        # All of these are initialized with the dummy nodes created.
        before = before_head = ListNode(0)
        after = after_head = ListNode(0)
        curr=head
        while curr:
            # If the original list node is lesser than the given x,
            # assign it to the before list.
            if curr.val < x:
                before.next = curr
                before = before.next
            else:
                # If the original list node is greater or equal to the given x,
                # assign it to the after list.
                after.next = curr
                after = after.next

            # move ahead in the original list
            curr = curr.next

        # Last node of "after" list would also be ending node of the reformed list
        after.next = None
        # Once all the nodes are correctly assigned to the two lists,
        # combine them to form a single list which would be returned.
        before.next = after_head.next

        return before_head.next
     
# @lc code=end

